src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -110,12 +118,62 @@ protected Object readResolve() throws ObjectStreamException {
      * It will also be false for something that has finished but is pending child node creation,
      * such as a completed fork branch which is waiting for the join node to be created.
      * <p>This can only go from true to false and is a shortcut for {@link FlowExecution#isCurrentHead}.
+     * @see #isActive
      */
     @Exported
     public final boolean isRunning() {
         return getExecution().isCurrentHead(this);
     }
 
+    private static final Map<FlowExecution, Set<BlockStartNode>> unclosedBlocks = new WeakHashMap<>();
+    @Restricted(DoNotUse.class)
+    @Extension public static class BlockListener implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            if (node instanceof BlockStartNode || node instanceof BlockEndNode) {
+                FlowExecution exec = node.getExecution();
+                synchronized (unclosedBlocks) {
```

Comment:
Ehhhh, any chance we could find a way to avoid all flows with BlockStartNodes and BlockEndNodes contending for this lock?   Maybe make a cache per FlowExecution or something?